source/solvers/tracer.cc

```diff
@@ -530,6 +553,179 @@ Tracer<dim>::calculate_tracer_statistics()
+                  fe_face_values_fd[velocities].get_function_values(
+                    current_solution_fd, velocity_values);
+
+                  // Loop on the quadrature points
+                  for (unsigned int q = 0; q < n_q_points_face; q++)
+                    {
+                      Tensor<1, dim> normal_vector_tracer =
+                        -fe_face_values_tracer.normal_vector(q);
+
+                      tracer_flow_rate_vector[vector_index] +=
+                        (-tracer_diffusivity[q] * tracer_gradient[q] *
+                           normal_vector_tracer +
+                         tracer_values[q] * velocity_values[q] *
+                           normal_vector_tracer) *
+                        fe_face_values_tracer.JxW(q);
+                    } // end loop on quadrature points
+                }     // end face is a boundary face
+            }         // end loop on faces
+        }             // end condition cell at boundary
+    }                 // end loop on cells
+
+
+  // Sum across all cores
+  for (unsigned int i_bc = 0;
+       i_bc < this->simulation_parameters.boundary_conditions.size;
+       ++i_bc)
+    tracer_flow_rate_vector[i_bc] =
+      Utilities::MPI::sum(tracer_flow_rate_vector[i_bc], mpi_communicator);
+
+  // Fill table
+  this->tracer_flow_rate_table.add_value(
+    "time", this->simulation_control->get_current_time());
+  this->tracer_flow_rate_table.set_precision("time", 12);
+  for (unsigned int i_bc = 0;
+       i_bc < this->simulation_parameters.boundary_conditions.size;
+       ++i_bc)
+    {
+      this->tracer_flow_rate_table.add_value("flow-rate-" +
+                                               Utilities::int_to_string(i_bc,
+                                                                        2),
+                                             tracer_flow_rate_vector[i_bc]);
+      this->tracer_flow_rate_table.set_scientific(
+        "flow-rate-" + Utilities::int_to_string(i_bc, 2), true);
+    }
+
+  // Console output
+  if (simulation_parameters.post_processing.verbosity ==
+      Parameters::Verbosity::verbose)
+    {
+      this->pcout << "Tracer flow rate at the boundaries: " << std::endl;
+      for (unsigned int i_bc = 0;
+           i_bc < this->simulation_parameters.boundary_conditions.size;
+           ++i_bc)
+        this->pcout << "\t boundary " << i_bc << ": "
+                    << tracer_flow_rate_vector[i_bc] << std::endl;
+    }
+
+  if ((simulation_control->get_step_number() %
+         this->simulation_parameters.post_processing.output_frequency ==
+       0) &&
+      Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+    {
+      std::string filename =
+        simulation_parameters.simulation_control.output_folder +
+        simulation_parameters.post_processing.tracer_flow_rate_output_name +
+        ".dat";
+      std::ofstream output(filename.c_str());
+      this->tracer_flow_rate_table.write_text(output);
+    }
```

Comment:
Personally I would fill the table write the output outside of the function and not inside. The function calculates the flow rate, it does not really take care of anything else. This should be two functions. One two calculate and one to store the information.